### PR TITLE
MGMT-7153 adding vsphere controller

### PR DESCRIPTION
**Pull request by michaellevy101**

### CODE REVIEW

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+from test_infra.helper_classes.config import BaseClusterConfig
+from test_infra.helper_classes.config.vsphere_config import VSphereControllerConfig
+from test_infra.tools import terraform_utils
+from test_infra.utils.terraform_util import TerraformControllerUtil
+
+
+class VSphereController(NodeController):
+
+    def __init__(self, config: VSphereControllerConfig, cluster_config: BaseClusterConfig):
+        super().__init__(config, cluster_config)
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:19]: `NodeController` has no `__init__` function that gets `VSphereControllerConfig` and   `BaseClusterConfig`

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    def get_network_by_name(self, network_name: str) -> libvirt.virNetwork:
+        raise NotImplementedError
+
+    def destroy_network(self, network: libvirt.virNetwork):
+        raise NotImplementedError
+
+    def set_single_node_ip(self, ip):
+        raise NotImplementedError
+
+    def __get_vm(self, node_name):
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:145]: Why private and not protected?

File: discovery-infra/test_infra/controllers/node_controllers/node_controller.py [hunk 3]
```diff
@@ -12,7 +12,9 @@
 
 class NodeController(ABC):
 
-    def __init__(self, config: BaseNodeConfig, cluster_config: BaseClusterConfig):
+    T = TypeVar('T', bound=SupportsAbs[BaseNodeConfig])
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/node_controller.py:15]: Annotations supposed to make thing easier to read.
And why not just `BaseNodeConfig`?

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
+import time
+from builtins import list
+from typing import Tuple, List, Callable, Any
+
+import libvirt
+from pyVim import task
+from pyVim.connect import SmartConnect, Disconnect
+from pyVmomi import vim
+
+from test_infra.controllers.node_controllers.disk import Disk
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:10]: Import from the package after declaring it in the __init__ file.
You will be able to merge imports:

```suggestion
from test_infra.controllers.node_controllers import Disk, Node, NodeController ..
```

File: discovery-infra/tests/base_test.py [hunk 5]
```diff
@@ -72,6 +77,10 @@ def cluster_configuration(self) -> ClusterConfig:
     @pytest.fixture
     def controller(self, cluster_configuration: ClusterConfig,
                    prepare_controller_configuration: BaseNodeConfig) -> NodeController:
+
+        if isinstance(prepare_controller_configuration, VSphereControllerConfig):
```
Comment (eliorerz) [inline discovery-infra/tests/base_test.py:81]: Why not use `cluster_configuration.platform` instead?

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py (hunk 1 above)
Comment (michaellevy101) [reply discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:23]: It has changed before this PR, It's ok

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py (hunk 2 above)
Comment (michaellevy101) [reply discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:177]: It's should private as all the other methods

File: discovery-infra/test_infra/controllers/node_controllers/node_controller.py (hunk 3 above)
Comment (michaellevy101) [reply discovery-infra/test_infra/controllers/node_controllers/node_controller.py:15]: It's a generic type with constraint that the config object should be a subtype of BaseNodeConfig
This way the compiler knows the real type in the derived controller class like VSphereController

**Review by Lewis-ver1803 — APPROVED** [review]

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+class VSphereController(NodeController):
+
+    def __init__(self, config: VSphereControllerConfig, cluster_config: BaseClusterConfig):
+        super().__init__(config, cluster_config)
+        self.cluster_name = cluster_config.cluster_name.get()
+        folder = TerraformControllerUtil.create_folder(self.cluster_name, platform=cluster_config.platform)
+        self._tf = terraform_utils.TerraformUtils(working_dir=folder)
+
+    def prepare_nodes(self):
+        config = {**self._config.get_all(), **self._cluster_config.get_all(),
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:27]: One line is fine, It's less than 120 characters

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    def undefine_interface(self, node_name: str, mac: str):
+        raise NotImplementedError
+
+    def create_network(self, network_xml: str) -> libvirt.virNetwork:
+        raise NotImplementedError
+
+    def get_network_by_name(self, network_name: str) -> libvirt.virNetwork:
+        raise NotImplementedError
+
+    def destroy_network(self, network: libvirt.virNetwork):
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:169]: `libvirt.virNetwork`?

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+    def add_interface(self, node_name, network_name, target_interface: str) -> str:
+        raise NotImplementedError
+
+    def undefine_interface(self, node_name: str, mac: str):
+        raise NotImplementedError
+
+    def create_network(self, network_xml: str) -> libvirt.virNetwork:
+        raise NotImplementedError
+
+    def get_network_by_name(self, network_name: str) -> libvirt.virNetwork:
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:166]: `libvirt.virNetwork`?

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+    def attach_interface(self, node_name, network_xml: str) -> Tuple[libvirt.virNetwork, str]:
+        raise NotImplementedError
+
+    def add_interface(self, node_name, network_name, target_interface: str) -> str:
+        raise NotImplementedError
+
+    def undefine_interface(self, node_name: str, mac: str):
+        raise NotImplementedError
+
+    def create_network(self, network_xml: str) -> libvirt.virNetwork:
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:163]: `libvirt.virNetwork`?

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+    def set_cpu_cores(self, node_name: str, core_count: int) -> None:
+        raise NotImplementedError
+
+    def set_ram_kib(self, node_name: str, ram_kib: int) -> None:
+        raise NotImplementedError
+
+    def get_machine_cidr(self) -> str:
+        raise NotImplementedError
+
+    def attach_interface(self, node_name, network_xml: str) -> Tuple[libvirt.virNetwork, str]:
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:154]: `libvirt.virNetwork`?

Comment (eliorerz) [reply discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:154]: At the moment I think you need to remove all libvirt annotations form the `NodeController`.

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+        raise NotImplementedError
+
+    def set_single_node_ip(self, ip):
+        raise NotImplementedError
+
+    def __get_vm(self, node_name):
+        return next((vm for vm in self.__get_vms() if vm["attributes"]["name"] == node_name), None)
+
+    def __get_vms(self):
+        vms_object_type = self._tf.get_resources(resource_type="vsphere_virtual_machine")
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:179]: Change the annotations of `get_resources` to something more specific. 
At the moment it's just list but you addressing it as List[Dict[string, something]]

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+    def destroy_network(self, network: libvirt.virNetwork):
+        raise NotImplementedError
+
+    def set_single_node_ip(self, ip):
+        raise NotImplementedError
+
+    def __get_vm(self, node_name):
+        return next((vm for vm in self.__get_vms() if vm["attributes"]["name"] == node_name), None)
+
+    def __get_vms(self):
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:178]: Return type annotation is missing

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+        vms_object_type = self._tf.get_resources(resource_type="vsphere_virtual_machine")
+
+        if not vms_object_type:
+            return list()
+
+        return vms_object_type[0]["instances"]
+
+    def __run_on_vm(self, name, folder, action):
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:186]: Type annotation are missing

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+                break
+
+        container.Destroy()
+
+        if vm is None:
+            raise ValueError("Unable to locate VirtualMachine.")
+
+        vm_task = action(vm)
+
+        while vm_task.info.state not in [vim.TaskInfo.State.success,
```
Comment (eliorerz) [inline discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:207]: One line

File: discovery-infra/tests/base_test.py [hunk 15]
```diff
@@ -25,9 +25,11 @@
 from test_infra.controllers.node_controllers import NodeController
 from test_infra.controllers.node_controllers import TerraformController
 from test_infra.controllers.node_controllers.node import Node
+from test_infra.controllers.node_controllers.vsphere_controller import VSphereController
```
Comment (eliorerz) [inline discovery-infra/tests/base_test.py:28]:
```suggestion
from test_infra.controllers.node_controllers import NodeController, Node, TerraformController, VSphereController
```

File: discovery-infra/tests/base_test.py [hunk 16]
```diff
@@ -25,9 +25,11 @@
...
 from test_infra.controllers.proxy_controller.proxy_controller import ProxyController
 from test_infra.helper_classes.cluster import Cluster
 from test_infra.helper_classes.config.controller_config import BaseNodeConfig, global_variables
+from test_infra.helper_classes.config.vsphere_config import VSphereControllerConfig
```
Comment (eliorerz) [inline discovery-infra/tests/base_test.py:32]:
```suggestion
from test_infra.helper_classes.config import BaseNodeConfig, global_variables, VSphereControllerConfig
```

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py (hunk 7 above)
Comment (michaellevy101) [reply discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:169]: The issue is from the NodeController. will fix in the PR

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py (hunk 8 above)
Comment (michaellevy101) [reply discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:166]: The issue is from the NodeController. will fix later after making the base general

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py (hunk 9 above)
Comment (michaellevy101) [reply discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:163]: The issue is from the NodeController. will fix later after making the base general

File: discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py (hunk 10 above)
Comment (michaellevy101) [reply discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:154]: The issue is from the NodeController. will fix later after making the base general

**Review by eliorerz — APPROVED** [review]
LGTM

**Pull request merged**